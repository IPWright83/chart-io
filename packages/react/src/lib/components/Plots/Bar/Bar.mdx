import { Story, Canvas } from "@storybook/addon-docs/blocks";

# Bar Plots

There are two different Bar based components, these include:

-   `<Bar>` - allows us to define a bar plot on our chart
-   `<Bars>` - allows us to define multiple bar series, or a stacked bar plot

> **Note**: When you are using a Bar plot you may need help the chart by specifying the type of scale. You can do this by configuring your `<XAxis>` like so `<XAxis ... scaleType="band" />`

<Canvas>
    <Story id="xycharts-bar--basic" />
</Canvas>

## `<Bar>` Component

The `<Bar>` component should be used when you want to plot a single bar plot (horizontal) on a chart.

### Props

| Prop          | Type      | Default            | Note                                                                                                         |
| ------------- | --------- | ------------------ | ------------------------------------------------------------------------------------------------------------ |
| `color`       | `string`  | Derived from Theme | Override the colour of the series                                                                            |
| **`x`\***     | `string`  | null               | The key of the field that contains the value along the x-axis.                                               |
| **`y`\***     | `string`  | null               | The key of the field that contains the value along the y-axis.                                               |
| `interactive` | `boolean` | true               | Whether the plot should be interactive. Setting this to false will disable tooltips & markers from this plot |

### Visual Styling

Most of the configuration for a `<Bar>` is similar to other plots. Currently the only styling property avaliable is the colour option.

<Canvas>
    <Story id="xycharts-bar--color" />
</Canvas>

## `<Bars>` Component

The `<Bars>` component allows us to define multiple bar series. There are three ways to combine multiple bar plots, depending on the type of bar chart you wish to produce:

-   Ratio Bars
-   Stacked Bars
-   Grouped Bars

### Props

| Prop       | Type       | Default            | Note                                                              |
| ---------- | ---------- | ------------------ | ----------------------------------------------------------------- |
| `colors`   | `string[]` | Derived from Theme | Override the colours for the series                               |
| **`xs`\*** | `string[]` | null               | The keys of the fields that contains the values along the x-axis. |
| **`y`\***  | `string`   | null               | The key of the field that contains value along the y-axis.        |
| `stacked`  | `boolean`  | false              | Whether the plots should be stacked.                              |
| `grouped`  | `boolean`  | false              | Whether the plots should be grouped together.                     |

### Ratio Charts

One way of combining Bars plots is to display one over the top of the other. This is useful if you want to show one value as a ratio of another. Remember if you do this that you need to disable the default opacity on a bar.

<Canvas>
    <Story id="xycharts-bar--ratio" />
</Canvas>

### Stacked Charts

Another way to to combine bar plots is to stack them on top of each other. To stack bar plots you **must** use the `<Bars>` (plural) component and ensure that you set stacked field `<Bars ... stacked={true} />`.

The chart will attempt to warn you, but you should also set the `aggregate` field on your X Axis/Scale as otherwise the data will overflow of the top of the scale `<XAxis ... aggregate={true} />`.

<Canvas>
    <Story id="xycharts-bar--stacked" />
</Canvas>

### Grouped Charts

Another way to to combine bar plots is to group them together in multiple series. To group bar plots you **must** use the `<Bars>` (plural) component and ensure that you set grouped field `<Bars ... grouped={true} />`.

<Canvas>
    <Story id="xycharts-bar--grouped" />
</Canvas>
