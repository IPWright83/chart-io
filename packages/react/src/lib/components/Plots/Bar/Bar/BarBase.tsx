import * as d3 from "@chart-it/d3";
import type { IDatum, IEventPlotProps, INumericValue } from "@chart-it/types";
import type { Selection, Transition } from "@chart-it/d3";

import { useEffect, useState } from "react";
import { useSelector, useStore } from "react-redux";

import { chartSelectors, eventActions, IState } from "../../../../store";
import { ensureBandScale, ensureValuesAreUnique } from "../../../../utils";
import { useLegendItem, useRender } from "../../../../hooks";

import { getDropline } from "../getDropline";
import { renderCanvas } from "../../renderCanvas";
import { useTooltip } from "../useTooltip";

export interface IBarBaseProps extends IEventPlotProps {
    /**
     * This is an internally used function to allow the scatter plot to render to a virtual canvas
     */
    renderVirtualCanvas?: (update: Transition<Element, unknown, any, unknown>) => void;
}

/**
 * Represents a Bar Plot
 * @param  props       The set of React properties
 * @return             The Bar plot component
 */
export function BarBase({
    x,
    y,
    canvas,
    renderVirtualCanvas,
    color,
    scaleMode = "plot",
    showInLegend = true,
    interactive = true,
    onMouseOver,
    onMouseOut,
    onClick,
    layer,
}: IBarBaseProps) {
    const [focused, setFocused] = useState(null);
    const store = useStore();

    const data = useSelector((s: IState) => chartSelectors.data(s));
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const xScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, x, scaleMode));
    const yScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, y, scaleMode));
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const animationDuration = useSelector((s: IState) => chartSelectors.animationDuration(s));

    const fillColor = d3.color(`${color ?? theme.series.colors[0]}`);
    fillColor.opacity = theme.series.opacity;
    const strokeColor = theme.background;

    useLegendItem(x, "square", showInLegend, fillColor);
    const setTooltip = useTooltip(store.dispatch, y);

    // This useEffect handles mouseOver/mouseExit through the use of the `focused` value
    useEffect(() => {
        if (!focused) return;

        const selection = d3.select(focused.element).style("opacity", theme.series.selectedOpacity);
        const dropline = getDropline(selection, yScale, false);
        store.dispatch(eventActions.addDropline(dropline));

        // Clean up operations on exit
        return () => {
            selection.style("opacity", theme.series.opacity);
            store.dispatch(eventActions.removeDropline(dropline));
        };
    }, [store.dispatch, focused, yScale, theme.series.selectedOpacity]);

    useRender(() => {
        if (ensureBandScale(yScale, "Bar") === false) return null;
        ensureValuesAreUnique(data, y, "Bar");

        // D3 data join
        const join = d3
            .select(layer.current)
            .selectAll(".bar")
            .data(data, (d) => d[y]) as Selection<SVGRectElement, IDatum, Element, unknown>;

        // Exit bars
        join.exit().remove();

        // Enter bars
        const enter = join
            .enter()
            .append("rect")
            .attr("class", "bar")
            .attr("x", () => xScale.range()[0])
            // @ts-ignore: TODO: Need to work out casting
            .attr("y", (d) => yScale(d[y]))
            .attr("width", () => 0)
            // @ts-expect-error: scale.bandwidth() has already been protected against using ensureBandScale()
            .attr("height", () => yScale.bandwidth())
            .style("stroke", strokeColor.toString())
            .style("fill", () => fillColor.toString());

        // Update new and existing points
        const update = enter
            .merge(join)
            .on("mouseover", function (event, datum) {
                // istanbul ignore next
                if (!interactive) return;

                onMouseOver && onMouseOver(datum, this, event);
                setFocused({ element: this, event, datum });
                setTooltip({ datum, event, fillColors: [fillColor], xs: [x] });
            })
            .on("mouseout", function (event, datum) {
                // istanbul ignore next
                if (!interactive) return;

                onMouseOut && onMouseOut(datum, this, event);
                setFocused(null);
                setTooltip(null);
            })
            .on("click", function (event, datum) {
                // istanbul ignore next
                if (!interactive) return;

                onClick(datum, this, event);
            })
            .transition("position")
            .duration(animationDuration / 2)
            // @ts-ignore: How do we deal with the scale here? y is likely a string
            .attr("y", (d) => yScale(d[y]))
            // @ts-expect-error: scale.bandwidth() has already been protected against using ensureBandScale()
            .attr("height", () => yScale.bandwidth())
            .style("fill", () => fillColor.toString())
            // @ts-expect-error: Looks like the type defs are wrong missing named transitions
            .transition("width")
            .duration(animationDuration / 2)
            .delay(animationDuration / 2)
            .attr("x", () => xScale.range()[0])
            .attr("width", (d) => xScale(d[x] as INumericValue) - xScale.range()[0]);

        renderCanvas(canvas, renderVirtualCanvas, width, height, update);
    }, [
        x,
        y,
        data,
        canvas,
        renderVirtualCanvas,
        xScale,
        yScale,
        layer,
        animationDuration,
        onMouseOver,
        onMouseOut,
        onClick,
    ]);

    return null;
}
