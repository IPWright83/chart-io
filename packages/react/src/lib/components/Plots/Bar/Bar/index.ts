export * from "./Bar";
export * from "./BarBase";
