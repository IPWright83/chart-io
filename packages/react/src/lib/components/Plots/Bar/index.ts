export * from "./Bar";
export * from "./StackedBars";
export * from "./GroupedBars";
export * from "./Bars";
