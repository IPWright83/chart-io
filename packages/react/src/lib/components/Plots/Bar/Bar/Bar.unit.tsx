import { scaleBand, scaleLinear } from "@d3-chart/d3";
import React from "react";
import { toMatchImageSnapshot } from "jest-image-snapshot";

import { VIRTUAL_CANVAS_DEBOUNCE, VirtualCanvas } from "../../../VirtualCanvas";
import { Bar } from "./Bar";

expect.extend({ toMatchImageSnapshot });

import {
    actionsIncludes,
    getBuffer,
    renderChart,
    testMouseClick,
    testMouseExit,
    testMouseOver,
    wait,
} from "../../../../testUtils";

describe("Bar", () => {
    const expectedDatum = {
        y: "A",
        x: 5,
    };

    const data = [
        { y: "A", x: 5 },
        { y: "B", x: 10 },
    ];

    const scales = {
        y: scaleBand().domain(["A", "B"]).range([0, 100]),
        x: scaleLinear().domain([0, 20]).range([0, 100]),
    };

    describe("using SVG", () => {
        it("should render correctly", async () => {
            const { asFragment } = await renderChart({
                children: <Bar x="x" y="y" />,
                data,
                scales,
            });

            expect(asFragment()).toMatchSnapshot();
        });

        describe("should handle event", () => {
            it("mouseover correctly", async () => {
                const onMouseOver = jest.fn();

                const { container, store } = await renderChart({
                    children: <Bar x="x" y="y" onMouseOver={onMouseOver} />,
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                testMouseOver(container, "rect", onMouseOver, expectedDatum);

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                actionsIncludes(dispatchCalls, [
                    "CHART.ADD_LEGEND_ITEM",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                ]);
            });

            it("mouseexit correctly", async () => {
                const onMouseOut = jest.fn();

                const { container, store } = await renderChart({
                    children: <Bar x="x" y="y" onMouseOut={onMouseOut} />,
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                testMouseExit(container, "rect", onMouseOut, expectedDatum);

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                actionsIncludes(dispatchCalls, [
                    "CHART.ADD_LEGEND_ITEM",
                    // Mouseexit
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_DROPLINE",
                ]);
            });

            it("click correctly", async () => {
                const onClick = jest.fn();

                const { container, store } = await renderChart({
                    children: <Bar x="x" y="y" onClick={onClick} />,
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                testMouseClick(container, "rect", onClick, expectedDatum);
            });
        });

        describe("should skip rendering if", () => {
            it("there is no x scale avaliable", async () => {
                const { asFragment } = await renderChart({
                    children: <Bar x="x" y="y" />,
                    data,
                    scales: { y: scales.y },
                });

                expect(asFragment()).toMatchSnapshot();
            });

            it("there is no y scale avaliable", async () => {
                const { asFragment } = await renderChart({
                    children: <Bar x="x" y="y" />,
                    data,
                    scales: { x: scales.x },
                });

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });

    describe("using Canvas", () => {
        it("should render correctly", async () => {
            const { container } = await renderChart({
                children: (
                    <VirtualCanvas>
                        <Bar x="x" y="y" useCanvas={true} />
                    </VirtualCanvas>
                ),
                data,
                scales,
            });

            await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

            const canvasBuffer = getBuffer(container.querySelector(".canvas"));
            expect(canvasBuffer).toMatchImageSnapshot();

            const virtualCanvasBuffer = getBuffer(container.querySelector(".virtual-canvas"));
            expect(virtualCanvasBuffer).toMatchImageSnapshot();
        });

        describe("should handle event", () => {
            it("mouseover correctly", async () => {
                const onMouseOver = jest.fn();

                const { container, store } = await renderChart({
                    children: (
                        <VirtualCanvas>
                            <Bar x="x" y="y" onMouseOver={onMouseOver} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                await testMouseOver(container, ".virtual-canvas", onMouseOver, expectedDatum, {
                    bubbles: true,
                    pageX: 10,
                    pageY: 10,
                });

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                actionsIncludes(dispatchCalls, [
                    "CHART.ADD_LEGEND_ITEM",
                    "EVENT.MOUSE_MOVE",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                ]);
            });

            it("mouseexit correctly", async () => {
                const onMouseOut = jest.fn();

                const { container, store } = await renderChart({
                    children: (
                        <VirtualCanvas>
                            <Bar x="x" y="y" onMouseOut={onMouseOut} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                await testMouseExit(
                    container,
                    ".virtual-canvas",
                    onMouseOut,
                    expectedDatum,
                    {
                        pageX: 10,
                        pageY: 10,
                    },
                    {
                        pageX: 95,
                        pageY: 95,
                    }
                );

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                actionsIncludes(dispatchCalls, [
                    "CHART.ADD_LEGEND_ITEM",
                    // Mouseover
                    "EVENT.MOUSE_MOVE",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",

                    // Mouseexit
                    "EVENT.MOUSE_MOVE",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_DROPLINE",
                ]);
            });

            it("click correctly", async () => {
                const onClick = jest.fn();

                const { container, store } = await renderChart({
                    children: (
                        <VirtualCanvas>
                            <Bar x="x" y="y" onClick={onClick} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                await testMouseClick(container, ".virtual-canvas", onClick, expectedDatum, {
                    bubbles: true,
                    pageX: 10,
                    pageY: 10,
                });
            });
        });
    });
});
