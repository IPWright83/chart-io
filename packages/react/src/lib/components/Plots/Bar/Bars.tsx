import type { IPlotsProps } from "@d3-chart/types";

import React from "react";
import { useSelector } from "react-redux";

import { chartSelectors, IState } from "../../../store";

import { Bar } from "./Bar";
import { GroupedBar } from "./GroupedBar";
import { StackedBar } from "./StackedBar";

export interface IBarsProps extends Omit<IPlotsProps, "ys" | "x"> {
    /**
     * The set of x fields to use to access the data for each plot
     */
    xs: Array<string>;
    /**
     * Should the column plots be stacked based on the x-value?
     */
    stacked?: boolean;
    /**
     * Should the column plots be grouped based on the x-value?
     */
    grouped?: boolean;
}

/**
 * Represents a set of Bar Plots
 * @param  props       The set of React properties
 * @return             The Bar plot component
 */
export function Bars({ xs, colors, stacked = false, grouped = false, ...props }: IBarsProps) {
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const palette = colors || theme.series.colors;

    if (stacked && grouped) {
        throw new Error("Bar plots currently do not support both being stacked and grouped");
    }

    if (!stacked && !grouped) {
        throw new Error("Multiple bars plots must be either stacked or grouped");
    }

    if (stacked) {
        return <StackedBar xs={xs} colors={palette} {...props} />;
    }

    if (grouped) {
        return <GroupedBar xs={xs} colors={palette} {...props} />;
    }

    return (
        <React.Fragment>
            {xs.map((x, i) => (
                <Bar {...props} key={x} x={x} color={palette[i]} />
            ))}
        </React.Fragment>
    );
}

Bars.requiresVirtualCanvas = true;
