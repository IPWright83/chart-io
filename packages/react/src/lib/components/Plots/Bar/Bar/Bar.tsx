import React from "react";

import { withCanvas, withSVG, withXYPlot } from "../../../../hoc";

import { BarBase, IBarBaseProps } from "./BarBase";

export interface IBarProps extends Omit<IBarBaseProps, "layer" | "canvas"> {
    /**
     * Should Canvas be used instead of SVG?
     */
    useCanvas?: boolean;
}

const CanvasBar = withCanvas(withXYPlot<IBarProps>(BarBase), "plot bar");
const SVGBar = withSVG(withXYPlot<IBarProps>(BarBase), "plot bar");

/**
 * Represents a Bar plot
 * @param  useCanvas   Should Canvas be used instead of SVG?
 * @param  props       The set of React properties
 * @return             The Bar plot component
 */
export function Bar({ useCanvas = false, ...props }: IBarProps) {
    if (useCanvas) {
        return <CanvasBar {...props} />;
    }

    return <SVGBar {...props} />;
}

Bar.requiresVirtualCanvas = true;
