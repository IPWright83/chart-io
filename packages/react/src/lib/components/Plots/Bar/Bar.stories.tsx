import { themes } from "@chart-io/core";

import React from "react";

import { argTypes } from "../../../../storybook/argTypes";
import { sales_records_dataset } from "../../../../data/sales_records_dataset";
import { Bar } from "./Bar";
import { Bars } from "./Bars";
import { XYChart } from "../../XYChart";
import { XAxis, YAxis } from "../../Axis";
import { createSVGTest, createCanvasTest } from "../../../testUtils";

import { uniqBy } from "lodash";

const { width, height, margin, useCanvas, theme, color } = argTypes;

export default {
  title: "XYCharts/Bar",
  component: Bar,
  parameters: {
    docs: {
      transformSource: (src) => {
        src = src.replace(/data={\[.*?\]}/gs, "data={[ ...dataset ]}");
        src = src.replaceAll(/undefined,?/g, "");
        src = src.replace(/^\s*\n/gm, "");
        return src;
      },
    },
    chromatic: { delay: 300 },
  },
  argTypes: {
    useCanvas,
    width,
    height,
    theme,
    color,
    leftMargin: margin,
    rightMargin: margin,
    topMargin: margin,
    bottomMargin: margin,
    onClick: { action: "clicked" },
    onMouseOver: { action: "onMouseOver" },
    onMouseOut: { action: "onMouseOut" },
  },
};

const data = uniqBy(sales_records_dataset, (d) => d["Item Type"]);

const BarTemplate = (args) => (
  <XYChart
    data={data}
    plotMargin={{
      left: args.leftMargin,
      right: args.rightMargin,
      top: args.topMargin,
      bottom: args.bottomMargin,
    }}
    width={args.width}
    height={args.height}
    animationDuration={args.animationDuration}
    theme={args.theme}
    useCanvas={args.useCanvas}
    onClick={args.onClick}
    onMouseOver={args.onMouseOver}
    onMouseOut={args.onMouseOut}
  >
    <YAxis fields={[args.y]} scaleType="band" showGridlines={false} />
    <XAxis fields={[args.x, args.x2, args.x3]} />
    <Bar x={args.x} y={args.y} color={args.color} />
    {args.x2 && <Bar x={args.x2} y={args.y} color={args.color2} />}
  </XYChart>
);

const BarsTemplate = (args) => (
  <XYChart
    data={data}
    plotMargin={{
      left: args.leftMargin,
      right: args.rightMargin,
      top: args.topMargin,
      bottom: args.bottomMargin,
    }}
    width={args.width}
    height={args.height}
    animationDuration={args.animationDuration}
    useCanvas={args.useCanvas}
    theme={args.theme}
    onClick={args.onClick}
    onMouseOver={args.onMouseOver}
    onMouseOut={args.onMouseOut}
  >
    <YAxis fields={[args.y]} scaleType="band" showGridlines={false} />
    <XAxis fields={[args.x, args.x2, args.x3]} aggregate={args.stacked} />
    <Bars
      y={args.y}
      xs={[args.x, args.x2]}
      grouped={args.grouped}
      stacked={args.stacked}
    />
  </XYChart>
);

export const Basic = {
  name: "Basic Plot",
  render: BarTemplate,
  args: {
    useCanvas: false,
    width: 800,
    height: 500,
    animationDuration: 250,
    color: "#99C1DC",
    color2: "#fc998e",
    theme: themes.light,
    leftMargin: 120,
    rightMargin: 40,
    topMargin: 40,
    bottomMargin: 40,
    x: "Unit Price",
    y: "Item Type",
  },
  play: createSVGTest("rect.bar", { clientX: 107, clientY: 396 }),
};

export const Color = {
  name: "Custom Color",
  render: BarTemplate,
  args: {
    ...Basic.args,
    color: "orange",
  },
};

export const Canvas = {
  name: "Using Canvas",
  render: BarTemplate,
  args: {
    ...Basic.args,
    useCanvas: true,
  },
  play: createCanvasTest({ clientX: 245, clientY: 455 }),
};

export const Ratio = {
  name: "Ratio Bars",
  render: BarTemplate,
  args: {
    ...Basic.args,
    x2: "Unit Cost",
    theme: {
      ...themes.light,
      series: {
        ...themes.light.series,
        opacity: 1,
      },
    },
  },
};

export const Stacked = {
  name: "Stacked Bars",
  render: BarsTemplate,
  args: {
    ...Basic.args,
    x2: "Unit Cost",
    stacked: true,
  },
  play: createSVGTest("rect.bar", { clientX: 107, clientY: 396 }),
};

export const Grouped = {
  name: "Grouped Bars",
  render: BarsTemplate,
  args: {
    ...Basic.args,
    x2: "Unit Cost",
    grouped: true,
  },
  play: createSVGTest("rect.bar", { clientX: 107, clientY: 396 }),
};
