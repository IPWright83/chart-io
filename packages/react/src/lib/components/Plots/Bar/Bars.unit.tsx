import { d3 } from "@chart-io/core";
import { toMatchImageSnapshot } from "jest-image-snapshot";
import React from "react";

import { VIRTUAL_CANVAS_DEBOUNCE, VirtualCanvas } from "../../VirtualCanvas";
import { Bars } from "./Bars";

expect.extend({ toMatchImageSnapshot });

import { getBuffer, renderChart, wait } from "../../../testUtils";

describe("Bars", () => {
    const data = [
        { y: "A", x: 5, x2: 8 },
        { y: "B", x: 10, x2: 12 },
    ];
    const scales = {
        y: d3.scaleBand().domain(["A", "B"]).range([0, 100]),
        x: d3.scaleLinear().domain([0, 30]).range([0, 100]),
        x2: d3.scaleLinear().domain([0, 30]).range([0, 100]),
    };

    describe("Stacked", () => {
        describe("throws error", () => {
            beforeEach(() => {
                jest.spyOn(console, "error").mockImplementation(() => jest.fn());
            });

            afterEach(() => {
                jest.clearAllMocks();
            });

            it("when neither stacked nor grouped", async () => {
                await expect(
                    renderChart({
                        children: <Bars y="y" xs={["x", "x2"]} />,
                        data,
                        scales,
                    })
                ).rejects.toThrow("Multiple Bar plots must be either stacked or grouped");
            });

            it("when stacked and grouped", async () => {
                await expect(
                    renderChart({
                        children: <Bars y="y" xs={["x", "x2"]} stacked grouped />,
                        data,
                        scales,
                    })
                ).rejects.toThrow("Bar plots do not support both being stacked and grouped");
            });
        });

        describe("using SVG", () => {
            it("should render correctly", async () => {
                const { asFragment } = await renderChart({
                    children: <Bars y="y" xs={["x", "x2"]} stacked />,
                    data,
                    scales,
                });

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("using Canvas", () => {
            it("should render correctly", async () => {
                const { container } = await renderChart({
                    children: (
                        <VirtualCanvas>
                            <Bars y="y" xs={["x", "x2"]} useCanvas={true} stacked />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                const canvases = container.querySelectorAll(".canvas");
                expect(canvases.length).toBe(1);

                const canvasBuffer1 = getBuffer(canvases[0] as HTMLCanvasElement);
                expect(canvasBuffer1).toMatchImageSnapshot();

                const virtualCanvasBuffer = getBuffer(container.querySelector(".virtual-canvas"));
                expect(virtualCanvasBuffer).toMatchImageSnapshot();
            });
        });
    });

    describe("Grouped", () => {
        describe("using SVG", () => {
            it("should render correctly", async () => {
                const { asFragment } = await renderChart({
                    children: <Bars y="y" xs={["x", "x2"]} grouped />,
                    data,
                    scales,
                });

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("using Canvas", () => {
            it("should render correctly", async () => {
                const { container } = await renderChart({
                    children: (
                        <VirtualCanvas>
                            <Bars y="y" xs={["x", "x2"]} useCanvas={true} grouped />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                const canvases = container.querySelectorAll(".canvas");
                expect(canvases.length).toBe(1);

                const canvasBuffer1 = getBuffer(canvases[0] as HTMLCanvasElement);
                expect(canvasBuffer1).toMatchImageSnapshot();

                const virtualCanvasBuffer = getBuffer(container.querySelector(".virtual-canvas"));
                expect(virtualCanvasBuffer).toMatchImageSnapshot();
            });
        });
    });
});
