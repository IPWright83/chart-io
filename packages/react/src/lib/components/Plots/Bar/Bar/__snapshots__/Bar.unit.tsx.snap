// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar using SVG should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot bar"
      clip-path="url(#clip-path-test)"
    >
      <rect
        class="bar"
        height="50"
        style="stroke: #ffffff; fill: rgba(153, 193, 220, 0.7);"
        width="0"
        x="0"
        y="0"
      />
      <rect
        class="bar"
        height="50"
        style="stroke: #ffffff; fill: rgba(153, 193, 220, 0.7);"
        width="0"
        x="0"
        y="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Bar using SVG should skip rendering if there is no x scale avaliable 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot bar"
      clip-path="url(#clip-path-test)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Bar using SVG should skip rendering if there is no y scale avaliable 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot bar"
      clip-path="url(#clip-path-test)"
    />
  </svg>
</DocumentFragment>
`;
